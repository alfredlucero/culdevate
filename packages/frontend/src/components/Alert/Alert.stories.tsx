import React from "react";
import { storiesOf } from "@storybook/react";
import Alert from "./index";

const onHide = () => {
  console.log("Hide alert!");
};

storiesOf("Common/Alert", module)
  .add("Success", () => (
    <Alert isShowing={true} onHide={onHide} variant="success">
      Success Alert
    </Alert>
  ))
  .add("Info", () => (
    <Alert isShowing={true} onHide={onHide} variant="info">
      Info Alert
    </Alert>
  ))
  .add("Warning", () => (
    <Alert isShowing={true} onHide={onHide} variant="warning">
      Warning Alert
    </Alert>
  ))
  .add("Danger", () => (
    <Alert isShowing={true} onHide={onHide} variant="danger">
      Danger Alert
    </Alert>
  ));
