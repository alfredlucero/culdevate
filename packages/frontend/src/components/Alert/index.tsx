import React from "react";
import cn from "classnames";
import { CommonProps } from "../commonProps";

export interface AlertProps extends CommonProps {
  isShowing: boolean;
  variant: AlertVariant;
  onHide: () => void;
  children: React.ReactNode;
}

type AlertVariant = "success" | "info" | "warning" | "danger";

const alertBaseClasses = [
  "border-l-4",
  "p-4",
  "cursor-pointer",
  "rounded",
  "shadow",
  "flex",
  "justify-between",
  "break-words",
];
const alertSuccessTextClass = "text-green-700";
const alertSuccessClasses = [alertSuccessTextClass, "bg-green-100", "border-green-500"];
const alertInfoTextClass = "text-blue-700";
const alertInfoClasses = [alertInfoTextClass, "bg-blue-100", "border-blue-500"];
const alertWarningTextClass = "text-orange-700";
const alertWarningClasses = [alertWarningTextClass, "bg-orange-100", "border-orange-500"];
const alertDangerTextClass = "text-red-700";
const alertDangerClasses = [alertDangerTextClass, "bg-red-100", "border-red-500"];

const Alert: React.FC<AlertProps> = ({
  isShowing,
  variant,
  onHide,
  children,
  testId = "",
  className = "",
  ...passThroughProps
}) => {
  if (!isShowing) {
    return null;
  }

  return (
    <div
      className={`${cn(
        alertBaseClasses,
        ...(variant === "success" ? alertSuccessClasses : []),
        ...(variant === "info" ? alertInfoClasses : []),
        ...(variant === "warning" ? alertWarningClasses : []),
        ...(variant === "danger" ? alertDangerClasses : []),
      )} ${className}`}
      onClick={onHide}
      role="alert"
      {...(testId !== "" ? { "data-testid": testId } : {})}
      {...passThroughProps}
    >
      <div className="w-5/6">{children}</div>
      <div className="w-1/6 flex justify-end items-center">
        <svg
          className={cn("fill-current", "h-6", "w-6", {
            [alertSuccessTextClass]: variant === "success",
            [alertInfoTextClass]: variant === "info",
            [alertWarningTextClass]: variant === "warning",
            [alertDangerTextClass]: variant === "danger",
          })}
          role="button"
          xmlns="http://www.w3.org/2000/svg"
          viewBox="0 0 20 20"
        >
          <title>Close</title>
          <path d="M14.348 14.849a1.2 1.2 0 0 1-1.697 0L10 11.819l-2.651 3.029a1.2 1.2 0 1 1-1.697-1.697l2.758-3.15-2.759-3.152a1.2 1.2 0 1 1 1.697-1.697L10 8.183l2.651-3.031a1.2 1.2 0 1 1 1.697 1.697l-2.758 3.152 2.758 3.15a1.2 1.2 0 0 1 0 1.698z" />
        </svg>
      </div>
    </div>
  );
};

export default Alert;
