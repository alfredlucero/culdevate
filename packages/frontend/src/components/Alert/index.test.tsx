import React from "react";
import { render, fireEvent } from "@testing-library/react";
import Alert from "./index";

describe("<Alert />", () => {
  const alertTestId = "alertTestId";

  test("should render without error", () => {
    const { container } = render(
      <div>
        <Alert variant="success" onHide={() => {}} isShowing={true} testId={alertTestId} className="extra-alert-class">
          Success Alert
        </Alert>
        <Alert variant="warning" onHide={() => {}} isShowing={true} testId={alertTestId} className="extra-alert-class">
          Warning Alert
        </Alert>
        <Alert variant="info" onHide={() => {}} isShowing={true} testId={alertTestId} className="extra-alert-class">
          Info Alert
        </Alert>
        <Alert variant="danger" onHide={() => {}} isShowing={true} testId={alertTestId} className="extra-alert-class">
          Danger Alert
        </Alert>
      </div>,
    );

    expect(container).toMatchSnapshot();
  });

  test("should not exist when the alert is not showing", () => {
    const { queryByTestId } = render(
      <Alert variant="success" onHide={() => {}} isShowing={false} testId={alertTestId}>
        Success Alert
      </Alert>,
    );

    expect(queryByTestId(alertTestId)).toBeNull();
  });

  test("should call onHide callback when clicking the alert", () => {
    const onHideMock = jest.fn();
    const { getByTestId } = render(
      <Alert variant="success" onHide={onHideMock} isShowing={true} testId={alertTestId}>
        Success Alert
      </Alert>,
    );

    fireEvent.click(getByTestId(alertTestId));

    expect(onHideMock).toHaveBeenCalled();
  });
});
