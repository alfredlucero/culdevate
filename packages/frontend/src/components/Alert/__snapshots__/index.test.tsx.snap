// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Alert /> should render without error 1`] = `
<div>
  <div>
    <div
      class="border-l-4 p-4 cursor-pointer rounded shadow flex justify-between break-words text-green-700 bg-green-100 border-green-500 extra-alert-class"
      data-testid="alertTestId"
      role="alert"
    >
      <div
        class="w-5/6"
      >
        Success Alert
      </div>
      <div
        class="w-1/6 flex justify-end items-center"
      >
        <svg
          class="fill-current h-6 w-6 text-green-700"
          role="button"
          viewBox="0 0 20 20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <title>
            Close
          </title>
          <path
            d="M14.348 14.849a1.2 1.2 0 0 1-1.697 0L10 11.819l-2.651 3.029a1.2 1.2 0 1 1-1.697-1.697l2.758-3.15-2.759-3.152a1.2 1.2 0 1 1 1.697-1.697L10 8.183l2.651-3.031a1.2 1.2 0 1 1 1.697 1.697l-2.758 3.152 2.758 3.15a1.2 1.2 0 0 1 0 1.698z"
          />
        </svg>
      </div>
    </div>
    <div
      class="border-l-4 p-4 cursor-pointer rounded shadow flex justify-between break-words text-orange-700 bg-orange-100 border-orange-500 extra-alert-class"
      data-testid="alertTestId"
      role="alert"
    >
      <div
        class="w-5/6"
      >
        Warning Alert
      </div>
      <div
        class="w-1/6 flex justify-end items-center"
      >
        <svg
          class="fill-current h-6 w-6 text-orange-700"
          role="button"
          viewBox="0 0 20 20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <title>
            Close
          </title>
          <path
            d="M14.348 14.849a1.2 1.2 0 0 1-1.697 0L10 11.819l-2.651 3.029a1.2 1.2 0 1 1-1.697-1.697l2.758-3.15-2.759-3.152a1.2 1.2 0 1 1 1.697-1.697L10 8.183l2.651-3.031a1.2 1.2 0 1 1 1.697 1.697l-2.758 3.152 2.758 3.15a1.2 1.2 0 0 1 0 1.698z"
          />
        </svg>
      </div>
    </div>
    <div
      class="border-l-4 p-4 cursor-pointer rounded shadow flex justify-between break-words text-blue-700 bg-blue-100 border-blue-500 extra-alert-class"
      data-testid="alertTestId"
      role="alert"
    >
      <div
        class="w-5/6"
      >
        Info Alert
      </div>
      <div
        class="w-1/6 flex justify-end items-center"
      >
        <svg
          class="fill-current h-6 w-6 text-blue-700"
          role="button"
          viewBox="0 0 20 20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <title>
            Close
          </title>
          <path
            d="M14.348 14.849a1.2 1.2 0 0 1-1.697 0L10 11.819l-2.651 3.029a1.2 1.2 0 1 1-1.697-1.697l2.758-3.15-2.759-3.152a1.2 1.2 0 1 1 1.697-1.697L10 8.183l2.651-3.031a1.2 1.2 0 1 1 1.697 1.697l-2.758 3.152 2.758 3.15a1.2 1.2 0 0 1 0 1.698z"
          />
        </svg>
      </div>
    </div>
    <div
      class="border-l-4 p-4 cursor-pointer rounded shadow flex justify-between break-words text-red-700 bg-red-100 border-red-500 extra-alert-class"
      data-testid="alertTestId"
      role="alert"
    >
      <div
        class="w-5/6"
      >
        Danger Alert
      </div>
      <div
        class="w-1/6 flex justify-end items-center"
      >
        <svg
          class="fill-current h-6 w-6 text-red-700"
          role="button"
          viewBox="0 0 20 20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <title>
            Close
          </title>
          <path
            d="M14.348 14.849a1.2 1.2 0 0 1-1.697 0L10 11.819l-2.651 3.029a1.2 1.2 0 1 1-1.697-1.697l2.758-3.15-2.759-3.152a1.2 1.2 0 1 1 1.697-1.697L10 8.183l2.651-3.031a1.2 1.2 0 1 1 1.697 1.697l-2.758 3.152 2.758 3.15a1.2 1.2 0 0 1 0 1.698z"
          />
        </svg>
      </div>
    </div>
  </div>
</div>
`;
